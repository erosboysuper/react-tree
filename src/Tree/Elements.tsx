import styled from 'styled-components'
import { m } from 'framer-motion'

export const Element = styled(m.div)<{ currentTheme: string; borderTop: boolean; selected: boolean }>`
  padding: 8px 15px 8px 4px;
  min-height: 20px;
  min-width: 0;
  transition: all 0.2s ease-in-out;
  &:hover {
    background: ${(props) => props.theme._themes[props.currentTheme || 'dark'].hoverBg};
    color: ${(props) => props.theme._themes[props.currentTheme || 'dark'].hoverText};
  }

  ${({ theme, currentTheme, borderTop }) => (borderTop ? `border-top: 1px solid ${theme._themes[currentTheme || 'dark'].separator};` : '')}

  ${(props) =>
    props.selected
      ? `
    background-color: ${props.theme._themes[props.currentTheme || 'dark'].selectedBg};
    color: ${props.theme._themes[props.currentTheme || 'dark'].selectedText};
  `
      : ''}
`

export const Empty = styled(Element)<{ currentTheme: string }>`
  color: ${({ theme, currentTheme }) => theme._themes[currentTheme || 'dark'].accentText};
  background-color: ${({ theme, currentTheme }) => theme._themes[currentTheme || 'dark'].accentBg};
`
