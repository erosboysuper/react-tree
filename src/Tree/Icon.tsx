// @flow
import * as React from 'react'
import { m } from 'framer-motion'
import styled from 'styled-components'

interface ReactTreeIcon {
  size: string | 'default'
  currentTheme?: string
  defaultIcon?: boolean
  children?: React.ReactNode
}

interface ISizes {
  [key: string]: string
}

const sizes: ISizes = {
  xlarge: '24px',
  large: '20px',
  default: '16px',
  small: '14px',
  xsmall: '10px'
}

const IconWrapper = styled(m.div)<{ size: string }>`
  display: flex;
  align-items: center;
  justify-content: center;
  width: ${({ size }) => sizes[size]};
  height: ${({ size }) => sizes[size]};
  text-align: center;
  padding: 2px;
  overflow: hidden;

  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
`

const ReactTreeIconContainer = styled(IconWrapper)<{ currentTheme: string }>`
  svg {
    height: 100%;
    width: 100%;

    * {
      fill: ${({ theme, currentTheme }) => `${theme._themes[currentTheme || 'dark'].icon}`};
      stroke: ${({ theme, currentTheme }) => `${theme._themes[currentTheme || 'dark'].icon}`};
    }
  }
`

const Icon = React.forwardRef<HTMLDivElement, ReactTreeIcon>(({ size = 'default', currentTheme = 'dark', defaultIcon = false, children }, ref) => {
  return defaultIcon ? (
    <ReactTreeIconContainer ref={ref} currentTheme={currentTheme} size={size}>
      {children}
    </ReactTreeIconContainer>
  ) : (
    <IconWrapper ref={ref} size={size}>
      {children}
    </IconWrapper>
  )
})

const MotionIcon = m(Icon, { forwardMotionProps: true })

export default MotionIcon
