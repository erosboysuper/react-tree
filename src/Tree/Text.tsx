import styled from 'styled-components'

export const NodeText = styled.span`
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow-x: hidden;
`
