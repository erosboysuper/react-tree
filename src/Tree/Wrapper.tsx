// @flow

import * as React from 'react'
import styled from 'styled-components'
import { m } from 'framer-motion'

const _Wrapper = styled(m.div)<{ level: number }>`
  padding-left: calc(${({ level }) => `25px * ${level}`});
  min-height: 20px;
  display: flex;
  align-items: center;
  min-width: 0;
`

const Wrapper: React.FC<{ level: number }> = ({ level, children }) => {
  return <_Wrapper level={level}>{children}</_Wrapper>
}

Wrapper.defaultProps = {
  level: 0
}

export default Wrapper
