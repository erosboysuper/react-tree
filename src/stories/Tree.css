:root {
  font-size: 62.5%;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
}

html {
  height: 100%;
}

*,
*::before,
*::after {
  box-sizing: border-box;
}

body {
  height: inherit;
}

#root {
  height: 100%;
  /* display: flex;
  flex-direction: column; */
}

h1 {
  font-size: 3.4rem;
}

h2,
h3 {
  font-size: 2rem;
}

h4,
p,
span,
ul,
ol,
dl,
li,
em,
strong,
details,
summary {
  font-size: 1.7rem;
}

/*
* Custom examples
*/
.custom-node {
  position: relative;
  height: 32px;
  display: flex;
  align-items: center;
}

.custom-node:hover {
  background-color: hotpink;
}

.custom-node span::before {
  position: absolute;
  top: 2px;
  left: var(--icon-pos);
  font-size: 1.7rem;
  font-style: normal;
  content: 'üìÅ';
}

.custom-node.open span::before {
  font-size: 1.7rem;
  content: 'üìÇ';
}

.custom-node.selected {
  background-color: black;
  color: white;
}

.custom-leaf {
  position: relative;
  height: 32px;
  display: flex;
  align-items: center;
}

.custom-leaf:hover {
  background-color: gold;
}

.custom-leaf.selected {
  background-color: black;
  color: white;
}

.custom-leaf span:before {
  position: absolute;
  top: 2px;
  font-size: 1.7rem;
  font-style: normal;
  content: 'üóíÔ∏è';
  left: var(--icon-pos);
}
