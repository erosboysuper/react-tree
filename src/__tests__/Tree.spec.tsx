import React from 'react'
import { cleanup, render, screen, waitFor } from '@testing-library/react'
import Tree from '..'

afterEach(cleanup)

describe('React Tree core functions', () => {
  it('Renders a message when there is no data', async () => {
    const { container } = render(<Tree nodes={[]} />)
    expect(container).toMatchSnapshot()
    const message = await screen.findByText(/No Data ðŸ˜”/i)
    expect(message).toBeTruthy()
  })

  it('Renders a loader and nothing else when `isLoading` state is `true`', async () => {
    const { container } = render(<Tree isLoading nodes={[]} />)
    expect(container).toMatchSnapshot()
    const message = await screen.findByText(/Loading/i)
    expect(message).toBeTruthy()
  })

  it('Re-renders nodes when the data set changes', async () => {
    const { rerender } = render(<Tree nodes={[{ id: 1, parentId: null, label: 'TEST_NODE_1' }]} />)
    const firstRender = await screen.findByText(/TEST_NODE_1/)
    expect(firstRender).toBeTruthy()
    rerender(<Tree nodes={[{ id: 2, parentId: null, label: 'TEST_NODE_2' }]} />)
    const secondRender = await screen.findByText(/TEST_NODE_2/)
    expect(secondRender).toBeTruthy()
  })
})
