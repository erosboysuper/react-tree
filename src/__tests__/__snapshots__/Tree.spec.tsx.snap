// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React Tree core functions Renders a loader and nothing else when \`isLoading\` state is \`true\` 1`] = `
<div>
  <div
    class="sc-pNWdM skpEp"
  >
    <div
      class="sc-jrsJWt dIZFVe"
    >
      <div>
        Loading... ðŸŒ€
      </div>
    </div>
  </div>
</div>
`;

exports[`React Tree core functions Renders a message when there is no data 1`] = `
<div>
  <div
    class="sc-pNWdM skpEp"
  >
    <div>
      <span>
        No Data ðŸ˜”
      </span>
    </div>
  </div>
</div>
`;
